### ISSUE TITLE
Delete Endpoint action results in 404

### ISSUE THREAD

**Issue by joecasanova** [splunk-soar-connectors/sophosendpointprotection#7]
Splunk SOAR version: 5.5.0.108488
Sophos Endpoint Protection version: 1.1.0

The app works with other actions without issue but I cannot delete endpoints.  The API credentials being used have superadmin permissions within Sophos central.

Below is the console output from SOAR when trying to run the delete endpoint action - please let me know what other data/info I can provide to help work through this issue:

```
============ Environment Debug ============

  Interpreter: Python 3.9.9

  Installed dependencies:

	

===========================================

Loaded action execution configuration

SophosEndpointProtectionConnector :Connector class File: sophosendpointprotection_connector 

SophosEndpointProtectionConnector :Asset config validated 

SophosEndpointProtectionConnector :App config validated 

SophosEndpointProtectionConnector :required keys [

    "endpointid"

]

SophosEndpointProtectionConnector :Creating state file: /opt/phantom/local_data/app_states/f94bd431-cfab-466b-8c09-3ae5954165ba/27_state.json 

SophosEndpointProtectionConnector :Loaded state:  {

    "app_version": "1.1.0",

    "jwt_json": {

        "access_token": "<REDACTED>",

        "errorCode": "success",

        "expires_in": 3600,

        "message": "OK",

        "refresh_token": "<REDACTED>",

        "token_type": "bearer",

        "trackingId": "<REDACTED>"

    },

    "pt_json": {

        "id": "<REDACTED>",

        "idType": "tenant",

        "apiHosts": {

            "global": "https://api.central.sophos.com",

            "dataRegion": "https://api-us01.central.sophos.com"

        }

    }

}

SophosEndpointProtectionConnector :required keys [

    "endpointid"

]

SophosEndpointProtectionConnector :Current param: {

    "endpointid": "<REDACTED>"

}

SophosEndpointProtectionConnector :__lock_info: {

    "enabled": true,

    "default_name": 27,

    "data_path": null,

    "timeout": 1200,

    "concurrency": true

}

SophosEndpointProtectionConnector :lock_name 27

SophosEndpointProtectionConnector :action_id delete_endpoint

In action handler for: delete_endpoint

idType: tenant

Hitting the URL: https://api-us01.central.sophos.com/endpoint/v1/endpoints

Trying to fetch data from the endpoint

SophosEndpointProtectionConnector :Action . Request to: https://api-us01.central.sophos.com/endpoint/v1/endpoints. Elapsed: 0.353788 seconds 

Response in JSON: None

Fetching the JWT token

Hitting the URL for JWT token: https://id.sophos.com/api/v2/oauth2/token

SophosEndpointProtectionConnector :Action . Request to: https://id.sophos.com/api/v2/oauth2/token. Elapsed: 0.374071 seconds 

Response in JSON: {'access_token': '<REDACTED>', 'token_type': 'bearer', 'trackingId': '<REDACTED>'}

Return value: True

Saving to state

Got the token: <REDACTED>

Fetching the partner token

SophosEndpointProtectionConnector :Action . Request to: https://api.central.sophos.com/whoami/v1. Elapsed: 0.406027 seconds 

SophosEndpointProtectionConnector :Saving state:  {

    "app_version": "1.1.0",

    "jwt_json": {

        "access_token": "<REDACTED>",

        "errorCode": "success",

        "expires_in": 3600,

        "message": "OK",

        "refresh_token": "<REDACTED>",

        "token_type": "bearer",

        "trackingId": "<REDACTED>"

    },

    "pt_json": {

        "id": "<REDACTED>",

        "idType": "tenant",

        "apiHosts": {

            "global": "https://api.central.sophos.com",

            "dataRegion": "https://api-us01.central.sophos.com"

        }

    }

}

Got the partner token

SophosEndpointProtectionConnector :Action . Request to: https://api-us01.central.sophos.com/endpoint/v1/endpoints. Elapsed: 0.306353 seconds 

Response in JSON: None

SophosEndpointProtectionConnector :Action execution #1 () finished. Elapsed: 1.4563696384429932 seconds 

SophosEndpointProtectionConnector :All actions () finished. Elapsed: 1.456495761871338 seconds 

SophosEndpointProtectionConnector :Saving state:  {

    "app_version": "1.1.0",

    "jwt_json": {

        "access_token": "<REDACTED>",

        "errorCode": "success",

        "expires_in": 3600,

        "message": "OK",

        "refresh_token": "<REDACTED>",

        "token_type": "bearer",

        "trackingId": "<REDACTED>"

    },

    "pt_json": {

        "id": "<REDACTED>",

        "idType": "tenant",

        "apiHosts": {

            "global": "https://api.central.sophos.com",

            "dataRegion": "https://api-us01.central.sophos.com"

        }

    }

}

SophosEndpointProtectionConnector :Error occurred with param: {"endpointid": "<REDACTED>"} and debug_data:

["{'r_status_code': 404}", "{'r_text': '{\\n    \"error\": \"NotFound\",\\n    \"correlationId\": \"<REDACTED>\",\\n    \"requestId\": \"<REDACTED>\"\\n}'}", "{'r_headers': {'Date': 'Mon, 16 Jan 2023 19:01:16 GMT', 'Content-Type': 'application/json', 'Content-Length': '145', 'Connection': 'keep-alive', 'X-Correlation-ID': '<REDACTED>', 'X-Request-ID': '4a94790e-021b-4256-9ef5-035815183943'}}", "{'r_status_code': 200}", "{'r_text': '{\"access_token\": \"<REDACTED>\", \"errorCode\": \"success\", \"expires_in\": 3600, \"message\": \"OK\", \"refresh_token\": \"<REDACTED>\", \"token_type\": \"bearer\", \"trackingId\": \"<REDACTED>\"}'}", "{'r_headers': {'Content-Type': 'application/json', 'Content-Length': '1502', 'Connection': 'keep-alive', 'Content-Security-Policy': \"default-src https:; script-src https: 'unsafe-inline'; style-src https: 'unsafe-inline'\", 'Date': 'Mon, 16 Jan 2023 19:01:16 GMT', 'Referrer-Policy': 'same-origin', 'Server': 'SophosID', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'SAMEORIGIN', 'X-XSS-Protection': '1; mode=block', 'X-Cache': 'Miss from cloudfront', 'Via': '1.1 fb7d525e4df8699605096cdd05f56b1c.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'DFW56-P2', 'X-Amz-Cf-Id': 'MzQq8vIU6mEC54YlNFTvSP_bZFLHOrUjuWgaQStALZtytOuvBYey4A=='}}", "{'r_status_code': 200}", "{'r_text': '{\"id\":\"<REDACTED>\",\"idType\":\"tenant\",\"apiHosts\":{\"global\":\"https://api.central.sophos.com\",\"dataRegion\":\"https://api-us01.central.sophos.com\"}}'}", "{'r_headers': {'Date': 'Mon, 16 Jan 2023 19:01:16 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'X-Correlation-ID': 'd6ad5afa-f5b6-48fd-aacc-696fb3ae17e6', 'X-Request-ID': '10c0f3dc-118a-4ebf-99ef-8d78b8d7882f'}}", "{'r_status_code': 404}", "{'r_text': '{\\n    \"error\": \"NotFound\",\\n    \"correlationId\": \"88cfb422-b162-4318-a1a1-1d60f4495dd4\",\\n    \"requestId\": \"074b966c-9ba5-4c2e-9c74-6cda0c7c5903\"\\n}'}", "{'r_headers': {'Date': 'Mon, 16 Jan 2023 19:01:17 GMT', 'Content-Type': 'application/json', 'Content-Length': '145', 'Connection': 'keep-alive', 'X-Correlation-ID': '<REDACTED>', 'X-Request-ID': '<REDACTED>'}}"] 

{"identifier": "delete_endpoint", "result_data": [{"data": [], "extra_data": [], "summary": {}, "status": "failed", "message": "Error from server. Status Code: 404 Data from server: {\n    \"error\": \"NotFound\",\n    \"correlationId\": \"88cfb422-b162-4318-a1a1-1d60f4495dd4\",\n    \"requestId\": \"<REDACTED>\"\n}", "parameter": {"endpointid": "<REDACTED>"}, "context": {}}], "result_summary": {"total_objects": 1, "total_objects_successful": 0}, "status": "failed", "message": "1 action failed", "exception_occured": false, "action_cancelled": false}

Traceback (most recent call last):

  File "/opt/phantom/apps/sophosendpointprotection_<REDACTED>/<REDACTED>_2023_01_16_19_01_15.py", line 32, in <module>

    raise Exception("Action Failed")

Exception: Action Failed
```

**Comment by edro15** [comment]
Hi @joecasanova ,

Thanks for reaching out!

From the debug messages you reported I suspect you're calling the action `delete item` instead of `delete endpoint`. You should hit URL : `https://api-us01.central.sophos.com/endpoint/v1/endpoints/<Endpoint ID>`. 

```
[...]
SophosEndpointProtectionConnector :action_id delete_item

In action handler for: delete_item

idType: tenant

Hitting the URL: https://api-us01.central.sophos.com/endpoint/v1/settings/blocked-items/<REDACTED>

Trying to fetch data from the endpoint
[...]
```

Could you please try to call the action `delete endpoint` by providing the opportune `endpoint id`? Please let us know if you still run into errors. Thanks for your support.

**Comment by edro15** [comment]
Hi @joecasanova ,

I see you edited the description of this issue. Now the action `delete endpoint` is called, but you still get an error. 
Unfortunately I don't have an environment to reproduce what you're reporting. However, if you have the chance to modify the code of the app in your environment, please apply the following change and verify whether fixes the error.

```diff
# sophosendpointprotection_connector.py@355,356 
- ret_val, response = self._make_rest_call_helper(action_result, endpoint, params=params,
+ ret_val, response = self._make_rest_call_helper(action_result, final_endpoint, params=params,
```

Hope it helps.

**Comment by joecasanova** [comment]
Success!  Please let me know if I can provide any additional feedback or information.  We just stood up this app in our SOAR platform and I'd be absolutely willing to help test any other bug fixes/development issues/etc related to this app.

```
============ Environment Debug ============

  Interpreter: Python 3.9.9

  Installed dependencies:

	

===========================================

Loaded action execution configuration

SophosEndpointProtectionConnector :Connector class File: sophosendpointprotection_connector 

SophosEndpointProtectionConnector :Asset config validated 

SophosEndpointProtectionConnector :App config validated 

SophosEndpointProtectionConnector :required keys [

    "endpointid"

]

SophosEndpointProtectionConnector :Creating state file: /opt/phantom/local_data/app_states/59f16ee4-8755-43ad-ba30-f1e67df63a2d/41_state.json 

SophosEndpointProtectionConnector :Loaded state:  {

    "app_version": "1.0.0",

    "jwt_json": {

        "access_token": "<REDACTED>",

        "errorCode": "success",

        "expires_in": 3600,

        "message": "OK",

        "refresh_token": "<REDACTED>",

        "token_type": "bearer",

        "trackingId": "<REDACTED>"

    },

    "pt_json": {

        "id": "<REDACTED>",

        "idType": "tenant",

        "apiHosts": {

            "global": "https://api.central.sophos.com",

            "dataRegion": "https://api-us01.central.sophos.com"

        }

    }

}

SophosEndpointProtectionConnector :required keys [

    "endpointid"

]

SophosEndpointProtectionConnector :Current param: {

    "endpointid": "<REDACTED>"

}

SophosEndpointProtectionConnector :__lock_info: {

    "enabled": true,

    "default_name": 41,

    "data_path": null,

    "timeout": 1200,

    "concurrency": true

}

SophosEndpointProtectionConnector :lock_name 41

SophosEndpointProtectionConnector :action_id delete_endpoint

In action handler for: delete_endpoint

idType: tenant

Hitting the URL: https://api-us01.central.sophos.com/endpoint/v1/endpoints/<REDACTED>

Trying to fetch data from the endpoint

SophosEndpointProtectionConnector :Action . Request to: https://api-us01.central.sophos.com/endpoint/v1/endpoints/<REDACTED>. Elapsed: 2.12374 seconds 

Response in JSON: {'deleted': True}

Response in JSON: {'deleted': True}

SophosEndpointProtectionConnector :Action execution #1 () finished. Elapsed: 2.127915382385254 seconds 

SophosEndpointProtectionConnector :All actions () finished. Elapsed: 2.1280484199523926 seconds 

SophosEndpointProtectionConnector :Saving state:  {

    "app_version": "1.0.0",

    "jwt_json": {

        "access_token": "<REDACTED>",

        "errorCode": "success",

        "expires_in": 3600,

        "message": "OK",

        "refresh_token": "<REDACTED>",

        "token_type": "bearer",

        "trackingId": "<REDACTED>"

    },

    "pt_json": {

        "id": "<REDACTED>",

        "idType": "tenant",

        "apiHosts": {

            "global": "https://api.central.sophos.com",

            "dataRegion": "https://api-us01.central.sophos.com"

        }

    }

}

{"identifier": "delete_endpoint", "result_data": [{"data": [{"endpoint_deleted": true}], "extra_data": [], "summary": {"total_endpoints_deleted": 1}, "status": "success", "message": "Total endpoints deleted: 1", "parameter": {"endpointid": "<REDACTED>"}, "context": {}}], "result_summary": {"total_objects": 1, "total_objects_successful": 1}, "status": "success", "message": "1 action succeeded", "exception_occured": false, "action_cancelled": false}
```

**Comment by edro15** [comment]
Hi @joecasanova ,

Apologies for the late reply. The issue you reported has been fixed in version 1.2.0 of the app. 🥳 Please, do not hesitate to open new issues whenever needed. Thanks.
